snuba/datasets/sessions.py

```diff
@@ -0,0 +1,265 @@
+    def process_message(self, message, metadata=None) -> Optional[ProcessedMessage]:
+        if message["duration"] is None:
+            duration = None
+        else:
+            duration = _collapse_uint32(int(message["duration"] * 1000))
+
+        # since duration is not nullable, the max duration means no duration
+        if duration is None:
+            duration = MAX_UINT32
+
+        processed = {
+            "session_id": str(uuid.UUID(message["session_id"])),
+            "distinct_id": str(uuid.UUID(message.get("distinct_id") or NIL_UUID)),
+            "seq": message["seq"],
+            "deleted": 0,
+            "org_id": message["org_id"],
+            "project_id": message["project_id"],
+            "retention_days": message["retention_days"],
+            "duration": duration,
+            "status": STATUS_MAPPING[message["status"]],
+            "errors": _collapse_uint16(message["errors"]) or 0,
+            "received": _ensure_valid_date(
+                datetime.utcfromtimestamp(message["received"])
+            ),
+            "started": _ensure_valid_date(
+                datetime.utcfromtimestamp(message["started"])
+            ),
+            "release": message.get("release"),
+            "environment": message.get("environment"),
+        }
+        return ProcessedMessage(action=ProcessorAction.INSERT, data=[processed])
+
+
+class SessionsDataset(TimeSeriesDataset):
+    def __init__(self) -> None:
+        all_columns = ColumnSet(
+            [
+                ("session_id", UUID()),
+                ("distinct_id", UUID()),
+                ("seq", UInt(64)),
+                ("org_id", UInt(64)),
+                ("project_id", UInt(64)),
+                ("retention_days", UInt(16)),
+                ("duration", UInt(32)),
+                ("status", UInt(8)),
+                ("errors", UInt(16)),
+                # note here that deleted is really only an indicator so we can
+                # figure out which values can be discarded.  Deleting of values
+                # does not update the materializations.
+                ("deleted", UInt(8)),
+                ("received", DateTime()),
+                ("started", DateTime()),
+                ("release", LowCardinality(Nullable(String()))),
+                ("environment", LowCardinality(Nullable(String()))),
+            ]
+        )
+
+        write_schema = ReplacingMergeTreeSchema(
+            columns=all_columns,
+            local_table_name=WRITE_LOCAL_TABLE_NAME,
+            dist_table_name=WRITE_DIST_TABLE_NAME,
+            prewhere_candidates=["project_id", "org_id"],
+            order_by="(org_id, project_id, toDate(started), session_id, cityHash64(toString(session_id)))",
+            partition_by="(toMonday(received))",
+            version_column="seq",
+            sample_expr="cityHash64(toString(session_id))",
+            settings={"index_granularity": 16384},
+        )
+
+        read_columns = ColumnSet(
```

Comment:
Please add a TODO to eventually expose the same schema as raw table and turn the query into aggregate during query processing since we cannot yet support it properly now, but it should be done this way so we can use both tables for queries.